### PR TITLE
Zoom

### LINKED ISSUE #1044: Revamp the Upload screen
As stated in [#1037](https://github.com/commons-app/apps-android-commons/issues/1037#issuecomment-354669983):

> It's quite odd to show the image being uploaded in the background. It seems to be cluttering up with the text in the screen (the text Wikimedia Commons policies is barely visible). It would be better to show a small preview of the image and let the users see it fully if they want to.

screenshot_2018-01-01-20-20-16

### PR DESCRIPTION
**Pull request by tanvidadu**
## Description

Fixes #1044 

Added the functionality to pinch zoom in upload activity. 

## Screenshots showing what changed
 
zoomdemo

### CODE REVIEW

**Comment by tanvidadu** [comment]
Sorry for delay !

File: app/build.gradle [hunk 1]
```diff
@@ -10,6 +10,7 @@ dependencies {
     implementation 'com.github.nicolas-raoul:Quadtree:ac16ea8035bf07'
     implementation 'fr.avianey.com.viewpagerindicator:library:2.4.1.1@aar'
     implementation 'in.yuvi:http.fluent:1.3'
+    implementation 'com.github.chrisbanes:PhotoView:2.0.0'
```
Comment (nicolas-raoul) [inline app/build.gradle]: Apache license so no problem :-)

Comment (tanvidadu) [reply app/build.gradle]: I tried to implement the pinch zoom on my own earlier but it had a flaw. While dragging the image the bounds weren't restricted as a result user was able to drag the image out of screen. Debugging it was becoming a tedious task so i used this to add functionality of pinch zoom :)

**Comment by neslihanturan** [comment]
Works great, however there are some minor issues @tanvidadu 
- Conflicts
- First click opens zoom as expected, however to get rid of zoom screen we need to click that button again. Since FAB icon remains still, it is hard to find how to close zoom screen (I found it happens when clicking same button again). Maybe you can change the icon of FAB on click to make user feel that that button is also way to close.

**Comment by tanvidadu** [comment]
@neslihanturan I will start working on the suggested changes 👍

**Comment by tanvidadu** [comment]
@neslihanturan I have made the changes accordingly and resolved the merged conflicts.
 I have added screenshots for the same.

zoom_out
zoom_in

**Comment by neslihanturan** [comment]
@tanvidadu I have tested the buttons, works perfect. One more issue:
- When I try to upload an image from my DSLR camera it doesn't display zoom screen since image resolution is too high (I guess) once it gave me this error with crash:
```
FATAL EXCEPTION: main
                                                                         Process: fr.free.nrw.commons.debug, PID: 5261
                                                                         java.lang.OutOfMemoryError
                                                                             at android.graphics.BitmapFactory.nativeDecodeStream(Native Method)
                                                                             at android.graphics.BitmapFactory.decodeStreamInternal(BitmapFactory.java:635)
                                                                             at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:611)
                                                                             at android.graphics.BitmapFactory.decodeResourceStream(BitmapFactory.java:444)
                                                                             at android.graphics.drawable.Drawable.createFromResourceStream(Drawable.java:840)
                                                                             at android.graphics.drawable.Drawable.createFromStream(Drawable.java:791)
                                                                             at android.widget.ImageView.resolveUri(ImageView.java:660)
                                                                             at android.widget.ImageView.setImageURI(ImageView.java:399)
                                                                             at com.github.chrisbanes.photoview.PhotoView.setImageURI(PhotoView.java:119)
                                                                             at fr.free.nrw.commons.upload.ShareActivity.zoomImageFromThumb(ShareActivity.java:757)
                                                                             at fr.free.nrw.commons.upload.ShareActivity.access$200(ShareActivity.java:86)
                                                                             at fr.free.nrw.commons.upload.ShareActivity$1.onClick(ShareActivity.java:284)
                                                                             at android.view.View.performClick(View.java:4463)
                                                                             at android.view.View$PerformClick.run(View.java:18770)
                                                                             at android.os.Handler.handleCallback(Handler.java:808)
                                                                             at android.os.Handler.dispatchMessage(Handler.java:103)
                                                                             at android.os.Looper.loop(Looper.java:193)
                                                                             at android.app.ActivityThread.main(ActivityThread.java:5292)
                                                                             at java.lang.reflect.Method.invokeNative(Native Method)
                                                                             at java.lang.reflect.Method.invoke(Method.java:515)
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
                                                                             at dalvik.system.NativeStart.main(Native Method)

```

But it works perfect with photos from phone.

**Comment by tanvidadu** [comment]
@neslihanturan I have compressed the high resolution images. I have tested the above fix on Samsung j7 max

**Comment by neslihanturan** [comment]
It crashes:
- Use this feature
- Go back to contributions activity
- Use this feature again

```
 Process: fr.free.nrw.commons.debug, PID: 4297
                                                                         java.lang.NullPointerException
                                                                             at android.graphics.Bitmap.createBitmap(Bitmap.java:741)
                                                                             at android.graphics.Bitmap.createScaledBitmap(Bitmap.java:602)
                                                                             at fr.free.nrw.commons.upload.ShareActivity.zoomImageFromThumb(ShareActivity.java:774)
                                                                             at fr.free.nrw.commons.upload.ShareActivity.access$200(ShareActivity.java:89)
                                                                             at fr.free.nrw.commons.upload.ShareActivity$1.onClick(ShareActivity.java:287)
```

**Comment by tanvidadu** [comment]
I tried the following steps but was unable to reproduce the error. 
With help of error log I was able to identify the reason for Null pointer exception. It is while doing compression. I have added a fix for it please let me know whether it's working or not.

**Comment by neslihanturan** [comment]
Here is a better log:
zoom_crash

**Comment by tanvidadu** [comment]
@neslihanturan , Thanks :)  for the log. I have fixed the error.

**Comment by neslihanturan** [comment]
I get this crash while testing:
```
 FATAL EXCEPTION: main
                                                                         Process: fr.free.nrw.commons.debug, PID: 5318
                                                                         java.lang.NoSuchMethodError: java.lang.Long.min
                                                                             at fr.free.nrw.commons.upload.ShareActivity.zoomImageFromThumb(ShareActivity.java:786)
                                                                             at fr.free.nrw.commons.upload.ShareActivity.access$200(ShareActivity.java:93)
                                                                             at fr.free.nrw.commons.upload.ShareActivity$1.onClick(ShareActivity.java:291)
                                                                             at android.view.View.performClick(View.java:4463)
                                                                             at android.view.View$PerformClick.run(View.java:18770)
                                                                             at android.os.Handler.handleCallback(Handler.java:808)
                                                                             at android.os.Handler.dispatchMessage(Handler.java:103)
                                                                             at android.os.Looper.loop(Looper.java:193)
                                                                             at android.app.ActivityThread.main(ActivityThread.java:5292)
                                                                             at java.lang.reflect.Method.invokeNative(Native Method)
                                                                             at java.lang.reflect.Method.invoke(Method.java:515)
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
          
```                                                                   at dalvik.system.NativeStart.main(Native Method)

**Comment by tanvidadu** [comment]
@neslihanturan i have fixed the issue !

**Comment by neslihanturan** [comment]
Works great! Do you have any idea why tests fails @tanvidadu ?

**Comment by tanvidadu** [comment]
It has something to do with this log , but i am unable to debug it : 

>  com.android.ide.common.process.ProcessException: Error while executing process /usr/local/android-sdk/build-tools/26.0.2/aapt with arguments {package -f --no-crunch -I /usr/local/android-sdk/platforms/android-26/android.jar -M /home/travis/build/commons-app/apps-android-commons/app/build/intermediates/manifests/full/beta/release/AndroidManifest.xml -S /home/travis/build/commons-app/apps-android-commons/app/build/intermediates/res/merged/beta/release -m -J /home/travis/build/commons-app/apps-android-commons/app/build/generated/source/r/beta/release -F /home/travis/build/commons-app/apps-android-commons/app/build/intermediates/res/beta/release/resources-betaRelease.ap_ -D /home/travis/build/commons-app/apps-android-commons/app/build/intermediates/multi-dex/beta/release/manifest_keep.txt --custom-package fr.free.nrw.commons -0 apk --output-text-symbols /home/travis/build/commons-app/apps-android-commons/app/build/intermediates/symbols/beta/release --no-version-vectors}

**Comment by tanvidadu** [comment]
@neslihanturan  is there anything that can be done to solve this error?

**Comment by misaochan** [comment]
I restarted the Travis build, lets see if it works. :)

Also, when was the last time you rebased on master?

**Comment by tanvidadu** [comment]
4 days ago :)

**Comment by misaochan** [comment]
Hmm. @maskaravivek do you think we can just ignore Travis this time? The failure seems to be technical from what I can see.

**Comment by maskaravivek** [comment]
I don't think we can ignore the build failure. The `:app:processBetaReleaseResources` task is failing that points to some resource issue. PFA screenshot of logs pointing to the resource causing issues. 

screen shot 2018-04-23 at 1 15 55 am

**Comment by tanvidadu** [comment]
@maskaravivek  @misaochan  thanks for helping in debugging the issue. I changed the location of resources, this should resolve the issue.

**Comment by tanvidadu** [comment]
How do we rerun the travis ?

**Comment by maskaravivek** [comment]
Both checks are passing now. Going ahead and merging it. :)

**Pull request merged**